catalog/transformer/catalog-transformer-csv-common/src/test/java/ddf/catalog/transformer/csv/common/CsvTransformerTest.java

```diff
@@ -0,0 +1,223 @@
+import com.google.common.collect.Sets;
+import ddf.catalog.data.Attribute;
+import ddf.catalog.data.AttributeDescriptor;
+import ddf.catalog.data.AttributeType;
+import ddf.catalog.data.Metacard;
+import ddf.catalog.data.MetacardType;
+import ddf.catalog.data.impl.AttributeImpl;
+import ddf.catalog.data.impl.BasicTypes;
+import ddf.catalog.data.impl.MetacardImpl;
+import ddf.catalog.data.impl.MetacardTypeImpl;
+import ddf.catalog.transform.CatalogTransformerException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Scanner;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.junit.Before;
+import org.junit.Test;
+
+public class CsvTransformerTest {
+
+  private static final List<AttributeDescriptor> ATTRIBUTE_DESCRIPTOR_LIST = new ArrayList<>();
+  private static Map<String, Attribute> metacardDataMap = new HashMap<>();
+
+  private static List<Metacard> metacardList = new ArrayList<>();
+  private static final int METACARD_COUNT = 2;
+
+  private static final String CSV_REGEX = "[\\n\\r,]";
+
+  private static final Object[][] ATTRIBUTE_DATA = {
+    {"attribute1", "value1", BasicTypes.STRING_TYPE},
+    {"attribute2", "value2", BasicTypes.STRING_TYPE},
+    {"attribute3", 101, BasicTypes.INTEGER_TYPE},
+    {"attribute4", 3.14159, BasicTypes.DOUBLE_TYPE},
+    {"attribute5", "value,5", BasicTypes.STRING_TYPE},
+    {"attribute6", "value6", BasicTypes.STRING_TYPE},
+    {"attribute7", "OBJECT", BasicTypes.OBJECT_TYPE},
+    {"attribute8", "BINARY", BasicTypes.BINARY_TYPE}
+  };
+
+  @Before
+  public void setup() {
+    ATTRIBUTE_DESCRIPTOR_LIST.clear();
+    metacardList.clear();
+    buildMetacardDataMap();
+    buildMetacardList();
+  }
+
+  @Test
+  public void getAllAttributes() {
+    Set<String> hiddenFields = new HashSet<>();
+    hiddenFields.add("attribute1");
+
+    Set<AttributeDescriptor> allAttributes =
+        CsvTransformer.getAllAttributes(metacardList, hiddenFields);
+    assertThat(allAttributes, hasSize(5));
+
+    Set<String> allAttributeNames =
+        allAttributes.stream().map(AttributeDescriptor::getName).collect(Collectors.toSet());
+
+    final Set<String> expectedAttributes =
+        Sets.newHashSet("attribute2", "attribute3", "attribute4", "attribute5", "attribute6");
+
+    assertThat(allAttributeNames, is(expectedAttributes));
+  }
+
+  @Test
+  public void getOnlyRequestedAttributes() {
+    Set<String> requestedAttributes = new HashSet<>();
+    requestedAttributes.add("attribute1");
+
+    Set<AttributeDescriptor> onlyRequestedAttributes =
+        CsvTransformer.getOnlyRequestedAttributes(
+            metacardList, requestedAttributes, Collections.emptySet());
+
+    assertThat(onlyRequestedAttributes, hasSize(1));
+    final String name = onlyRequestedAttributes.stream().findFirst().get().getName();
```

Comment:
This should make an assertion that the result of `findFirst()` is present before calling `get()` and checking the value inside.

Reply:
Normally I don't check my `get` calls with `findFirst` in tests because if the optional is not present my test should fail anyway.

Reply:
The main difference is the ability for the test to display more useful, descriptive error messages when assertions are used instead of just waiting for Exceptions.